src/Utilities/cachingoptimizer.jl

```diff
@@ -208,25 +220,27 @@ function MOI.supportsconstraint(m::CachingOptimizer, F::Type{<:MOI.AbstractFunct
     MOI.supportsconstraint(m.model_cache, F, S) && (m.state == NoOptimizer || MOI.supportsconstraint(m.optimizer, F, S))
 end
 
-function MOI.canaddconstraint(m::CachingOptimizer, ::Type{F}, ::Type{S}) where {F<:MOI.AbstractFunction, S<:MOI.AbstractSet}
-    MOI.canaddconstraint(m.model_cache, F, S) || return false
-    if m.state == AttachedOptimizer && m.mode == Manual
-        MOI.canaddconstraint(m.optimizer, F, S) || return false
-    end
-    return true
-end
-
 function MOI.addconstraint!(m::CachingOptimizer, func::MOI.AbstractFunction, set::MOI.AbstractSet)
-    # The canaddconstraint checks should catch most issues, but if an
-    # addconstraint! call fails then the model_cache and the optimizer may no longer
-    # be in sync.
-    if m.mode == Automatic && m.state == AttachedOptimizer && !MOI.canaddconstraint(m.optimizer, typeof(func), typeof(set))
-        resetoptimizer!(m)
+    if m.state == AttachedOptimizer
+        if m.mode == Automatic
+            try
+                cindex_optimizer = MOI.addconstraint!(m.optimizer, mapvariables(m.model_to_optimizer_map, func), set)
+            catch err
+                if err isa MOI.CannotAddConstraint
+                    # It could be MOI.CannotAddConstraint{F', S'} with F' != F
```

Comment:
I'm not sure I understand. If `F` and `S` don't match, that seems like a leak of an internal implementation detail, like http://www.juliaopt.org/JuMP.jl/latest/style.html#User-facing-MethodError-1.

Reply:
There is no way around that, the bridges only check that the constraints created are supported but they do not check if it can be added since `canaddconstraint` do not exists anymore. Using `try` `catch` in bridges just for transforming the error doesn't seem like a good idea and it would hide the initial error.
This is unlikely to happen anyway in JuMP since bridges are added on top of the caching optimizer so the user will need to manually add bridges below to caching optimizer for this to happen in JuMP.

Reply:
I'm not entirely convinced on this point, maybe I don't understand bridges well enough. But this is relatively minor, we can address it later.